Reformatting nested_map function

RabeeBello opened the pull request:
Close #17420

RabeeBello commented:
add_reformatting_checklist_nest

KareemMAX reviewed (changes requested):
Hi @RabeeBello, great work

But in order to merge the PRs we need to at least show how other parameters work as well, you demosnterated well how `x` and `fn` works. You will need to show other parameters like `include_derived`, `to_ignore` etc..

You don't need to demonstrate it in a combinatorial way though, you may just use only lists for example for all other parameters

Final note on linting, we have added documentation lint checks recently, you can find what your PR is failing in [checks](https://github.com/unifyai/ivy/actions/runs/5370598990/jobs/9742790952?pr=17502#step:5:111), I recommend you install `pre-commit` to make sure your code is always linted before committing

Thanks

KareemMAX commented:
Hey @RabeeBello, noticed you are pushing commits to your PR.

Just a friendly reminder to unset your PR as draft when you finish and request a review from me

Thanks 😄

RabeeBello commented:
Hey @KareemMAX , I will surely do those things when I am done. Still working on it now :)

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,65 @@ def nested_map(
+    [['aH','bH','cH'],['dH','eH','fH'],['gH',['hH','iH']]]
+
+    >>> x = [[1, 2, 3, 4, 5], [23, 24, 25, 26, 27]]
+    >>> ivy.nested_map(x, include_derived = True)
```
I guess you forgot to add a function here

RabeeBello replied:
@KareemMAX Thank you for your corrections. I will add the functions where applicable. In terms of to_ignore and extra_nest_types, I am not entirely sure how to mention them, could please give me an idea ?

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,65 @@ def nested_map(
+    >>> print(nest)
+    [[3, 5, 7], [6, 9, 10]]
+
+    With :class:`ivy.Container` input:
+
+    >>> x = ivy.Container(a=ivy.array([1., 2.]) , b=ivy.array([4., 5.]))
+    >>> function = lambda _: ivy.array([3., 4.])
+    >>> ivy.nested_map(x, function)
+    >>> print(x)
+    {
+        a: ivy.array([3., 4.]),
+        b: ivy.array([3., 4.])
+    }
```
I think you forgot to reorder that. It would be better if it was with the other input types you mentioned above

RabeeBello replied:
@KareemMAX do you mean it is not necessary to show an 'ivy container' instance for the 'nested_map' function?

KareemMAX replied:
Nope, I meant that I guess it was suppose to be placed higher up. With `Tuple` and other containers.

KareemMAX reviewed (changes requested):
Hey @RabeeBello, nice work

I've some few comments, please look into them. Also I think you forgot to mention `to_ignore` and `extra_nest_types`

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,82 @@ def nested_map(
+    [[24, 25, 1338], [64, 99, 7]]
+
+    >>> nest = ([3, 5, 7], [6, 9, 10])
+    >>> ivy.nested_map(nest, to_mutable = True, shallow = False)
```
I can't understand what this argument does

KareemMAX reviewed (changes requested):
Hi @RabeeBello

I guess we need to have clarifications on the examples you wrote, because they doesn't show what the arguments are doing. Think as if you were the user reading this documentation, how would you like it to better understand the function.

Also don't forget to bring `ivy.Container` up to be below `List` and `Tuple` and other container like structures. I don't think it belongs down in its place now

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+    [[2, 4], (6, 8), {'a': 5, 'b': 6}]
+
+    Note:
+        - The function should be a callable that takes one argument.
```
This note is quite well known, isn't it? I guess it should be removed

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+
+    Note:
+        - The function should be a callable that takes one argument.
+        - The iterable should be a nested iterable, such as a list of lists.
```
Is that true? I believe it will work on lists as a well

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+    Note:
+        - The function should be a callable that takes one argument.
+        - The iterable should be a nested iterable, such as a list of lists.
+        - The to_ignore parameter allows specifying types to be ignored when deciding
+          whether to go deeper into the nest or not.
```
This is redundant, we already explained that while explaining the parameters.

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+        - The iterable should be a nested iterable, such as a list of lists.
+        - The to_ignore parameter allows specifying types to be ignored when deciding
+          whether to go deeper into the nest or not.
+        - By default, no types are ignored, and all nested types are recursively mapped.
```
This note is quite well known, isn't it? I guess it should be removed

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+        - The to_ignore parameter allows specifying types to be ignored when deciding
+          whether to go deeper into the nest or not.
+        - By default, no types are ignored, and all nested types are recursively mapped.
+        - Providing types in to_ignore prevents the function from recursively
+          mapping those types and stops the recursion at that level.
```
This is redundant, we already explained that while explaining the parameters.

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+
+    >>> nest = [[1, 2], {'a': [3, 4]}, 'hello']
+    >>> ivy.nested_map(lambda x: x * 2, nest, extra_nest_types=(dict,))
+    [[2, 4], {'a': [6, 8]}, 'hello']
```
If I understand correctly, you are trying to show adding extra nest types. `dict` is nest-able by default and this doesn't show any new functionality

RabeeBello replied:
@KareemMAX Do I have to add examples for all parameters ? I am finding it very confusing especially for Include_derived, extra_nest_types and max_depth. I do not understand what exactly you want me to show, could you please explain further? Will the other examples that that I have done that are fine suffice? Can I remove these other parameters as I did with shallow as well?

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+    >>> ivy.nested_map(lambda x: x * 2, nest, extra_nest_types=(dict,))
+    [[2, 4], {'a': [6, 8]}, 'hello']
+
+    Note:
+        - The function should be a callable that takes one argument.
+        - The iterable should be a nested iterable, such as a list of lists.
+        - The extra_nest_types parameter allows specifying additional types to
+          recursively check when deciding whether to go deeper into the nest or not.
+        - By default, only the base types (list, tuple, and dict) are
+          recursively checked.
+        - Providing extra types allows customizing the recursive behavior for
+          specific types not handled by default.
```
Those notes are quite similar to the above mentioned ones. I guess you should remove them as well

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+        - Providing extra types allows customizing the recursive behavior for
+          specific types not handled by default.
+
+    >>> nest = [[1, 2], (3, 4), {'a': 5, 'b': 6}]
+    >>> ivy.nested_map(lambda x: x * 2, nest, include_derived=True)
+    [[2, 4], (6, 8), {'a': 10, 'b': 12}]
```
Can't see anything resampling a derived type here

KareemMAX replied:
I'm not really aware here what derived types we can show, if you think there is a small example to show it off, please add it. Otherwise you can remove it

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+        - By default, the mapping is limited to the base types (list, tuple, dict)
+          and does not apply to derived classes.
+
+    >>> ivy.nested_map(lambda x: x * 2, [[1, 2], [3, 4], [5, 6]], max_depth=1)
```
This list have only 1 level

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+        - If max_depth is set to an integer value, the mapping stops when the
+          maximum depth is reached.
+
+    >>> ivy.nested_map(lambda x: x * 2, [[1, 2], [3, 4], [5, 6]], to_mutable=True)
```
In the last iteration you where using `Tuple` that worked well with this example, don't you think you should present it instead?

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,119 @@ def nested_map(
+        - By default, the nested iterable is returned as an immutable tuple.
+
+    >>> nest = [[1, 2], [3, 4], [5, 6]]
+    >>> ivy.nested_map(lambda x: x * 2, nest, shallow=True)
```
I'm thinking this parameter is a bit hard to show, you can remove it if you like. But this example isn't really showing any added functionality

RabeeBello replied:
I have removed the example I have for the shallow parameter.

KareemMAX reviewed (changes requested):
Hey @RabeeBello

Still I see that you didn't provide examples showing the functionality of parameters. Also, I find most of the added notes redundant

KareemMAX reviewed (changes requested):
Hey @RabeeBello,

Seems like I didn't asses the complexity of the parameters enough 😅. So you can remove some which I mentioned below.

Thanks

RabeeBello commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,68 @@ def nested_map(
+
+    >>> nest = [[1, 2], (3, 4), {'a': 5, 'b': 6}]
+    >>> ivy.nested_map(lambda x: x * 2, nest, to_ignore=(list, dict))
+    [[2, 4], (6, 8), {'a': 5, 'b': 6}]
```
@KareemMAX  I have removed 'extra_nest_types' and 'include_derived'. I have also changed the max_depth list to a 3D list and corrected the output for the 'to_ignore'.

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+    []
+
+    >>> nest = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]]
+    >>> nested_map(lambda x: x * 2, nest, max_depth=2)
```
I guess you meant to keep `max_depth=1` to show off what it does

KareemMAX replied:
Error here as well

```python
>>> nest = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]]
>>> nested_map(lambda x: x * 2, nest, max_depth=2)
  File "<stdin>", line 1, in <module>

NameError: name 'nested_map' is not defined
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+
+    >>> nest = [[1, 2], (3, 4), {'a': 5, 'b': 6}]
+    >>> ivy.nested_map(lambda x: x * 2, nest, to_ignore=(list, dict))
+    []
```
Are you sure that is the actual output?

KareemMAX replied:
Actually this results an error

```python
>>> nest = [[1, 2], (3, 4), {'a': 5, 'b': 6}]
>>> ivy.nested_map(lambda x: x * 2, nest, to_ignore=(list, dict))
  File "<stdin>", line 1, in <module>
  File "/workspaces/ivy/ivy/utils/exceptions.py", line 246, in _handle_exceptions
    raise ivy.utils.exceptions.IvyBackendException(

IvyBackendException: numpy: nested_map:   File "/workspaces/ivy/ivy/utils/exceptions.py", line 210, in _handle_exceptions
    return fn(*args, **kwargs)
  File "/workspaces/ivy/ivy/functional/ivy/nest.py", line 1178, in nested_map
    return fn(x)
During the handling of the above exception, another exception occurred:
 'list' object is not callable
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+        )
+    >>> function = lambda a : a  + 1
+    >>> ivy.nested_map(x, function)
+    >>> print(x)
+    {
+        a: ivy.array([[2, 3, 4], [10, 8, 8]]),
+        b: ivy.array([[5, 6,  7], [13, 14, 15]])
+    }
```
`x` actually doesn't change. Seems like you didn't run that on your console

KareemMAX reviewed (changes requested):
Seems like the examples you wrote don't checkout with what actually happen in ivy. Can you fix that?

RabeeBello commented:
@KareemMAX Hello, So I decided to run docstring examples tests and I got the following error. <img width="743" alt="Screenshot 2023-07-13 at 5 53 12 PM" src="https://github.com/unifyai/ivy/assets/123169759/e7008c2b-dac6-44d1-9d00-5e35248f89ba">

KareemMAX reviewed (changes requested):
I think otherwise, I believe the output of the docstring tests is basically saying you are writing mismatched docs (which I proved they were mismatched manually).

Can you please fix the examples you wrote rather than removing them? Consult using a python console to run you examples

RabeeBello commented:
@KareemMAX  Hello, I have fixed the issues you pointed out earlier in my examples and I have undone their removal.

Regarding the docstring examples test I mentioned above, I used the code below to run it using my python console of course. (Code is in the link)
https://gist.github.com/hirwa-nshuti/90053c86d8a7b507028c93c3d12a7157

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+
+    >>> nest = [[1, 2], [3, 4], [5, 6]]
+    >>> ivy.nested_map(lambda x: x * 2, nest, to_ignore=(list, dict))
+    [[1, 2], [3, 4], [5, 6]]
```
That looks okay, but what do you think of adding types that are not in `to_ignore`? Maybe removing `dict` from `to_ignore` and adding it in the nested object.

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+
+    >>> nest = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]]
+    >>> ivy.nested_map(lambda x: x * 2, nest, max_depth=1)
+    [[[2, 4], [6, 8]], [[10, 12], [14, 16]]]
```
Again, I can't see what `max_depth` does as it seems like it applies the lambda function to the whole object. Can you experiment with that parameter a little bit to make show an example that the lambda function is partially applied somehow. Of course it should be an actual valid code with its output

KareemMAX reviewed (changes requested):
Looks awesome so far, we are really close to merging your PR @RabeeBello.

I've some comments on `to_ignore` and `max_depth` that need to be resolved. You might need to do some experiments with the function in your python console, you should copy the output of your code from it given it shows how the function works in a meaningful way.

Always put yourself in the shoes of an ivy user. If you read the docstring you editing would it help you understand what function is doing?

Also, thank you so much for sticking along. I know it has been quite a long review, but it's coming to an end very soon. Thank you so much for your efforts 🚀🚀

RabeeBello commented:
@KareemMAX Thank you for the words of encouragement. I have removed dict from 'to_ignore' as you suggested and changed the output. I have also changed the nest for 'max_depth' to a multilevel nest and fixed the output as well. I think it now shows what the 'max_depth' function does.

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+
+    >>> x = ([[1., 2.], [3., 4.]])
+    >>> function = lambda a : a * 2
+    >>> ivy.nested_map(x, function)
+    >>> print(x)
+    ([[2.0, 4.0], [6.0, 8.0]])
```
No need to use `print(x)`

```suggestion
    >>> ivy.nested_map(x, function)
    ([[2.0, 4.0], [6.0, 8.0]])
```

Although I know you might want to show that `x` have actually been changed, so this can also be written (tbh I prefer it more)
```suggestion
    >>> ivy.nested_map(x, function)
    ([[2.0, 4.0], [6.0, 8.0]])
    >>> print(x)
    ([[2.0, 4.0], [6.0, 8.0]])
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+
+    >>> x = {1 : [1, [2, 3]], 2: (4, 5)}
+    >>> function = lambda a : a + 1
+    >>> ivy.nested_map(x, function)
+    >>> print(x)
+    {1 : [2, [3, 4]], 2: (5, 6)}
```
Same here
```suggestion
    >>> ivy.nested_map(x, function)
    {1 : [2, [3, 4]], 2: (5, 6)}
```
or
```suggestion
    >>> ivy.nested_map(x, function)
    {1 : [2, [3, 4]], 2: (5, 6)}
    >>> print(x)
    {1 : [2, [3, 4]], 2: (5, 6)}
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+    ...      ['d', 'e', 'f'],
+    ...      ['g', ['h', 'i']]]
+    >>> function = lambda a: a + 'H'
+    >>> ivy.nested_map(x, function)
+    >>> print(x)
+    [['aH','bH','cH'],['dH','eH','fH'],['gH',['hH','iH']]]
```
```suggestion
    >>> ivy.nested_map(x, function)
    [['aH','bH','cH'],['dH','eH','fH'],['gH',['hH','iH']]]
```
or
```suggestion
    >>> ivy.nested_map(x, function)
    [['aH','bH','cH'],['dH','eH','fH'],['gH',['hH','iH']]]
    >>> print(x)
    [['aH','bH','cH'],['dH','eH','fH'],['gH',['hH','iH']]]
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+
+    With :class:`ivy.Container` input:
+
+    >>> x = ivy.Container(
+      a=ivy.array([[1, 2, 3], [9, 8, 7]]) , b=ivy.array([[4, 5, 6], [12, 13, 14]])
+    )
```
This is incorrect

```suggestion
    >>> x = ivy.Container(
    ...   a=ivy.array([[1, 2, 3], [9, 8, 7]]) , b=ivy.array([[4, 5, 6], [12, 13, 14]])
    ... )
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+      a=ivy.array([[1, 2, 3], [9, 8, 7]]) , b=ivy.array([[4, 5, 6], [12, 13, 14]])
+    )
+    >>> function = lambda a : a  + 1
+    >>> ivy.nested_map(x, function)
+    >>> print(x)
+    {
+       a: ivy.array([[1, 2, 3], [9, 8, 7]]),
+       b: ivy.array([[4, 5,  6], [12, 13, 14]])
+    }
```
```suggestion
    >>> ivy.nested_map(x, function)
    {
       a: ivy.array([[1, 2, 3], [9, 8, 7]]),
       b: ivy.array([[4, 5,  6], [12, 13, 14]])
    }
```
or
```suggestion
    >>> ivy.nested_map(x, function)
    {
       a: ivy.array([[1, 2, 3], [9, 8, 7]]),
       b: ivy.array([[4, 5,  6], [12, 13, 14]])
    }
    >>> print(x)
    {
       a: ivy.array([[1, 2, 3], [9, 8, 7]]),
       b: ivy.array([[4, 5,  6], [12, 13, 14]])
    }
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+       b: ivy.array([[4, 5,  6], [12, 13, 14]])
+    }
+
+    >>> nest = [[1, 2], [3, 4], [5, 6], {"a": 1, "b": 2, "c": 3}]
+    >>> ivy.nested_map(lambda x: x * 2, nest, to_ignore=(list))
+    [[1, 2], [3, 4], [5, 6], {"a": 2, "b": 4, "c": 6}]
```
This code fragment is not working

```python
>>> nest = [[1, 2], [3, 4], [5, 6], {"a": 1, "b": 2, "c": 3}]
>>> ivy.nested_map(lambda x: x * 2, nest, to_ignore=(list))
Traceback (most recent call last):
  File "/workspaces/ivy/ivy/utils/exceptions.py", line 335, in _handle_exceptions
    return fn(*args, **kwargs)
  File "/workspaces/ivy/ivy/functional/ivy/nest.py", line 1187, in nested_map
    return fn(x)
TypeError: 'list' object is not callable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workspaces/ivy/ivy/utils/exceptions.py", line 372, in _handle_exceptions
    raise ivy.utils.exceptions.IvyBackendException(
ivy.utils.exceptions.IvyBackendException: numpy: nested_map: 'list' object is not callable
```

Please fix it

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+    >>> ivy.nested_map(lambda x: x * 2, nest, to_ignore=(list))
+    [[1, 2], [3, 4], [5, 6], {"a": 2, "b": 4, "c": 6}]
+
+    >>> nest = [[1, 2], [3, [4, 5]], [[6], [7, 8, [9, 10]]]]
+    >>> ivy.nested_map(lambda x: x * 2, nest, max_depth=1)
+    [[2, 4], [6, [4, 5]], [[6], [14, 16, [9, 10]]]]
```
This code fragment is not working as well

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+    [[2, 4], [6, [4, 5]], [[6], [14, 16, [9, 10]]]]
+
+    >>> nest = ([23, 25, 1337], [63, 98, 6])
+    >>> function = lamba a :  a + 1
```
```suggestion
    >>> function = lambda a :  a + 1
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,61 @@ def nested_map(
+
+    >>> nest = ([23, 25, 1337], [63, 98, 6])
+    >>> function = lamba a :  a + 1
+    >>> ivy.nested_map(nest, function, to_mutable = True)
+    >>> print(nest)
+    [[24, 25, 1338], [64, 99, 7]]
```
I guess using `print` would be redundant here, I'll only suggest this edit

```suggestion
    >>> ivy.nested_map(nest, function, to_mutable = True)
    [[24, 25, 1338], [64, 99, 7]]
```

KareemMAX reviewed (changes requested):
I've added some comments, some of which I'll apply myself and the others I'll leave them to you. For the demos that are not running correctly please be sure to **run the code you wrote before requesting a review**, and **don't transform the code fragments so they don't display what the respective arguments are doing**

RabeeBello commented:
@KareemMAX  Hello, I have fixed the code for both 'to_ignore' and 'max_depth'. It should work well now, however for 'max_depth' the default is 1 but you can increase it if your nest is deeper so for my example I increased it to 3 as I feel that it gives a clearer picture of what the function does. Thanks again for your patience.

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,69 @@ def nested_map(
+    >>> x = ([[1., 2.], [3., 4.]])
+    >>> function = lambda a : a * 2
+    >>> ivy.nested_map(x, function)
+    ([[2.0, 4.0], [6.0, 8.0]])
+    >>> print(x)
+    ([[2.0, 4.0], [6.0, 8.0]])
```
```suggestion
    [[2.0, 4.0], [6.0, 8.0]]
    >>> print(x)
    [[2.0, 4.0], [6.0, 8.0]]
```

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,69 @@ def nested_map(
+    ... )
+    >>> function = lambda a : a  + 1
+    >>> ivy.nested_map(x, function)
+    {
+       a: ivy.array([[1, 2, 3], [9, 8, 7]]),
+       b: ivy.array([[4, 5,  6], [12, 13, 14]])
+    }
+    >>> print(x)
+    {
+       a: ivy.array([[1, 2, 3], [9, 8, 7]]),
+       b: ivy.array([[4, 5,  6], [12, 13, 14]])
+    }
```
```suggestion
    {
        a: ivy.array([[2, 3, 4], 
                      [10, 9, 8]]),
        b: ivy.array([[5, 6, 7], 
                      [13, 14, 15]])
    }
    >>> print(x)
    {
        a: ivy.array([[2, 3, 4], 
                      [10, 9, 8]]),
        b: ivy.array([[5, 6, 7], 
                      [13, 14, 15]])
    }
```

KareemMAX reviewed (changes requested):
Hey @RabeeBello, great work.

Seems like you missed some things which I'll fix myself in the output of your examples

KareemMAX commented on ivy/functional/ivy/nest.py:
```diff
@@ -1055,6 +1055,73 @@ def nested_map(
+
+    >>> nest = ([1, 2], [3, 4], [5, 6], {"a": 1, "b": 2, "c": 3})
+    >>> function = lambda a :  a * 2
+    >>> ivy.nested_map(nest, function,  to_ignore = (list) )
```
```suggestion
    >>> ivy.nested_map(nest, function,  to_ignore=list)
```

KareemMAX commented:
ivy-gardener

KareemMAX reviewed (approved):
@RabeeBello your PR finally passed linting check and everything seems good. Thank you so much for your patience and effort 🚀 .

Merging your PR right now.
